src/InteractiveWindow/Editor/InteractiveWindow.UIThreadOnly.cs

```diff
@@ -631,7 +632,77 @@ public void AddInput(string command)
                 FinishCurrentSubmissionInput();
             }
 
-            private void AppendUncommittedInput(string text)
+            /// <summary>Implements <see cref="IInteractiveWindow2.AddToHistory"/>.</summary>
+            public void AddToHistory(string command) 
+            {
+                if (string.IsNullOrEmpty(command)) 
+                {
+                    return;
+                }
+
+
+                var historyBuffer = _factory.CreateAndActivateBuffer(_window);
+                _buffer.Flush();
+
+                using (var edit = historyBuffer.CreateEdit(EditOptions.None, null, s_suppressPromptInjectionTag)) 
+                {
+                    if (!command.EndsWith(_lineBreakString)) 
+                    {
+                        edit.Insert(0, command + _lineBreakString);
+                    } 
+                    else 
+                    {
+                        edit.Insert(0, command);
+                    }
+                    edit.Apply();
+                }
+                var end = historyBuffer.CurrentSnapshot.Length;
+
+                var historySpan = new SnapshotSpan(historyBuffer.CurrentSnapshot, Span.FromBounds(0, end -_lineBreakString.Length));
+                _history.Add(historySpan);
+
+                var sourceSpans = _projectionBuffer.CurrentSnapshot.GetSourceSpans();
+                int insertionIndex;
+                switch (State) 
+                {
+                    case State.Starting:
+                    case State.Initializing:
+                        throw new InvalidOperationException(InteractiveWindowResources.NotInitialized);
+                    case State.Resetting:
+                    case State.ResettingAndReadingStandardInput:
+                        throw new InvalidOperationException(InteractiveWindowResources.IsResettings);
+                    case State.ExecutingInput:
+                        insertionIndex = sourceSpans.Count;
+                        break;
+                    case State.ExecutingInputAndReadingStandardInput:
+                    case State.WaitingForInput:
+                    case State.WaitingForInputAndReadingStandardInput:
+                        insertionIndex = IndexOfLastSpan(sourceSpans, ReplSpanKind.Prompt);
+                        break;
+                    default:
+                        throw new ArgumentOutOfRangeException();
+                }
```

Comment:
I'd move this to the top of the method, before the history buffer is created and a span is added to _history.

That way we won't leave the state inconsistent in case of an exception thrown here